l5kit/l5kit/rasterization/box_rasterizer.py

```diff
@@ -55,14 +54,14 @@ def draw_boxes(
     else:
         im = np.zeros((raster_size[1], raster_size[0], 3), dtype=np.uint8)
 
-    box_world_coords = np.zeros((len(agents), 4, 2))
-    corners_base_coords = np.asarray([[-1, -1], [-1, 1], [1, 1], [1, -1]])
+    corners_base_coords = (np.asarray([[-1, -1], [-1, 1], [1, 1], [1, -1]]) / 2)[None, :, :]
 
     # compute the corner in world-space (start in origin, rotate and then translate)
-    for idx, agent in enumerate(agents):
-        corners = corners_base_coords * agent["extent"][:2] / 2  # corners in zero
-        r_m = yaw_as_rotation33(agent["yaw"])
-        box_world_coords[idx] = transform_points(corners, r_m) + agent["centroid"][:2]
+    corners_m = corners_base_coords * agents["extent"][:, None, :2]  # corners in zero
+    s = np.sin(agents["yaw"])
+    c = np.cos(agents["yaw"])
+    rotation_m = np.moveaxis(np.array(((c, s), (-s, c))), 2, 0)
```

Comment:
can we add a comment here saying that this is clockwise because it's right-multiplied and not left-multiplied later (and therefore we're still rotating conter-clockwise)?

Reply:
My understanding is that the corner base should be at +-0.5. So later when we multiply by extend, it will be at +-0.5*extend. But I guess they are equivalent.

Reply:
In a special case, you could also place your corners at -0.3*extent and 0.7*extent. In this case multiply extend by 0.5 or 0.3 will not work.

Reply:
> My understanding is that the corner base should be at +-0.5. So later when we multiply by extend, it will be at +-0.5*extend. But I guess they are equivalent.

They are equivalent. What you want to achieve is place the corners where they should be before rotating and translating



> Also, I think it is still faster to multiply 0.5 to corners_base_coords before it is exploded by the agent extent since after the multiplication to the `agent['extent']`, the matrix becomes (N agents) * (4 corners).

Yes, it should be faster. let leave it there then